### ISSUE TITLE
Script

### ISSUE THREAD

**Issue by Stargazer284** [Stargazer284/Bot-discord-per-le-piante#2]
import os

os.environ["TF_USE_LEGACY_KERAS"] = "1"

import io

import discord
import numpy as np
import tensorflow as tf
from PIL import Image
from discord import app_commands

TOKEN = os.environ["DISCORD_TOKEN"]

class CompatibleDepthwiseConv2D(tf.keras.layers.DepthwiseConv2D):
    def __init__(self, *args, groups=1, **kwargs):
        super().__init__(*args, **kwargs)

model = tf.keras.models.load_model(
    "keras_model.h5",
    compile=False,
    custom_objects={"DepthwiseConv2D": CompatibleDepthwiseConv2D},
)

with open("labels.txt", "r", encoding="utf-8") as file:
    labels = [line.strip() for line in file]

# Ogni indice corrisponde alla posizione della classe nel modello Teachable Machine.
informazioni_classi = {
    0: (
        "Taraxacum spp.",
        "I fiori forniscono nettare e polline, soprattutto all’inizio della primavera. "
        "I semi alimentano alcuni uccelli e le radici profonde contribuiscono ad aerare "
        "il terreno e a recuperare nutrienti dagli strati inferiori.",
    ),
    1: (
        "Trifolium repens; Trifolium pratense",
        "Sono importanti fonti di nettare per api e bombi. In simbiosi con batteri "
        "presenti nelle radici fissano l’azoto atmosferico, aumentando la fertilità "
        "del suolo e riducendo il bisogno di concimi azotati.",
    ),
    2: (
        "Urtica dioica",
        "È una pianta nutrice per i bruchi di numerose farfalle. Forma piccoli rifugi "
        "per insetti e altri animali e produce biomassa ricca di nutrienti, utile nel "
        "compostaggio.",
    ),
    3: (
        "Lamium purpureum",
        "Fiorisce già nei mesi freddi, offrendo nettare e polline quando le risorse sono "
        "ancora limitate. È visitata soprattutto da bombi e api dalla lingua lunga e "
        "protegge il terreno lasciato scoperto.",
    ),
    4: (
        "Achillea millefolium",
        "Le infiorescenze, formate da molti piccoli fiori, sono accessibili ad api, "
        "farfalle, sirfidi e insetti predatori o parassitoidi. Questi ultimi possono "
        "contribuire al controllo naturale degli organismi dannosi per le colture.",
    ),
    5: (
        "Cichorium intybus",
        "La lunga fioritura estiva offre polline e nettare anche durante periodi caldi "
        "e relativamente asciutti. Le radici profonde migliorano la struttura del "
        "terreno e la circolazione dell’acqua.",
    ),
    6: (
        "Echium vulgare",
        "Produce fiori ricchi di nettare, molto frequentati da api selvatiche, api "
        "domestiche, bombi e farfalle. È particolarmente utile negli ambienti aridi, "
        "nei terreni poveri e lungo i margini stradali.",
    ),
    7: (
        "Centaurea spp.",
        "Le fioriture estive forniscono abbondante nettare e polline a numerosi "
        "impollinatori. Dopo la fioritura, i semi possono essere consumati dagli "
        "uccelli. Alcune specie rappresentano inoltre piante nutrici per le larve di "
        "farfalle e falene.",
    ),
    8: (
        "Bellis perennis",
        "Nei prati poco falciati può fiorire per gran parte dell’anno, offrendo una "
        "fonte continua, sebbene modesta, di nettare e polline a piccoli insetti "
        "impollinatori. Tollera il calpestio e contribuisce alla copertura del suolo.",
    ),
    9: (
        "Plantago lanceolata; Plantago major",
        "Proteggono il terreno dall’erosione e resistono bene al calpestio. Il polline "
        "viene utilizzato da alcuni insetti, mentre i semi nutrono uccelli e piccoli "
        "animali. P. lanceolata è anche una pianta nutrice per i bruchi di alcune "
        "farfalle.",
    ),
}

client = discord.Client(intents=discord.Intents.default())
tree = app_commands.CommandTree(client)

@client.event
async def on_ready():
    await tree.sync()
    print(f"Bot collegato come {client.user}")

@tree.command(
    name="riconosci",
    description="Riconosce un'immagine con Teachable Machine",
)
async def riconosci(
    interaction: discord.Interaction,
    immagine: discord.Attachment,
):
    await interaction.response.defer()

    if not immagine.content_type or not immagine.content_type.startswith("image/"):
        await interaction.followup.send("Allega un file immagine.")
        return

    contenuto = await immagine.read()
    image = Image.open(io.BytesIO(contenuto)).convert("RGB")
    image = image.resize((224, 224))

    data = np.asarray(image, dtype=np.float32)
    data = (data / 127.5) - 1
    data = np.expand_dims(data, axis=0)

    prediction = model.predict(data, verbose=0)[0]
    index = int(np.argmax(prediction))
    percentuale = float(prediction[index]) * 100

    # Se l'indice non è presente nella mappatura, usa la label originale.
    nome, descrizione = informazioni_classi.get(
        index,
        (labels[index], "Nessuna informazione aggiuntiva disponibile."),
    )

    await interaction.followup.send(
        f"Risultato: ***{nome}***\n"
        f"Sicurezza: **{percentuale:.1f}%**\n\n"
        f"{descrizione}"
    )

client.run("TOKEN")
